src/rest.js

```diff
@@ -20,12 +20,23 @@ function connect(address, system) {
 
     connection.monitor = function(options, callback, system, return_raw) {
         return new Promise((resolve, reject) => {
+            let buffer = "";
+
             http.request(options)
                     .stream(data => {
                         if (return_raw)
                             callback(data);
-                        else
-                            callback(JSON.parse(data));
+                        else {
+                            buffer += data;
+
+                            try {
+                                const message = JSON.parse(buffer);
+                                buffer = "";
+                                callback(message);
+                            } catch {
+                                // Ignore, we already stored it is buffer
+                            }
```

Comment:
This is too brittle and probabilistic. Why not just let the bridge do the buffer assembly? I suggest using the original code, but as a .then() handler, instead of .stream(). We are not expecting huge replies here, and even if they were, the stream handler can't process individual chunks anyway -- so building up a buffer happens in any case. So let's just let cockpit.js do the job.

See https://cockpit-project.org/guide/latest/cockpit-http.html#cockpit-http-then

Reply:
We handle `.then()` as well (each user of `.monitor()` declares it on they own), but it means that the monitor has closed. We use monitor for streaming of events as they happen - we open the connection and sit on it for the whole duration of session. And we just get messages on what has happened.

Reply:
Ah, that's the monitor. So are we guaranteed to never get a chunk which includes the end of one JSON message and the start of another? Like this, if we'd get a first chunk like `{ a: "one", `, and a second one `"b": "two }\n{ "c": "three" }`, the first + second would either still be unparseable, or it would cut off and ignore the second object (as this resets `buffer`). If this is really getting a continuous stream of JSON blobs, then there must be *some* defined message separator? That's exactly what [jsonl](https://jsonlines.org/) is, after all. Then it could append new chunks to the buffer, check if the buffer contains `\n`, and if so, split it and process all parts but the last one (and keep that in the buffer).

Reply:
I believe it is guaranteed. We never saw that and the current behavior would be affected the same way as well.
The API does not try to send more messages in one reply, it just sometimes needs to split them.

Reply:
You can't really guarantee this with a TCP socket, this would require a SEQPACKET one. AFAICS the only reliable way (and best practice everywhere) is to use jsonl. This was literally invented for streaming json messages. I just ran `curl --unix-socket /run/user/1000/podman/podman.sock http://localhost/v1.12/libpod/events` and it looks correct -- this is proper jsonl.